### ISSUE TITLE
experiment(agent): compare Task 2 rehearsal and legal masking with escape inputs

### ISSUE THREAD

**Issue by 1BlauNitrox** [1BlauNitrox/mle-final-project#124]
## Resource-ledger recovery — September 10

Fixed the verified Windows `PermissionError` during atomic replacement of `resources.json` in commit `5a9d8ec`. The operational monitor retries only permission errors, up to 20 attempts with 0.1-second pauses. Persistent failures remain fatal. Validation: 746 tests and 11 subtests passed; lint and compilation passed.

The campaign resumed at 16:54 Berlin time under Python supervisor PID 38196. All completed training was preserved (A 100, B 63, C 95, D 55 blocks at handover). B/r2/block16 is now running attempt 003; its earlier interrupted/failed attempts remain retained. The retry input checkpoint checksum matches the retained block-15 checkpoint. A's completed training is skipped; B/C/D continue. Resource accounting remains cumulative and no limit breach is recorded.

The original learning source, protocol, 24-hour/96-CPU-hour ceiling, four-worker maximum and 8 GiB RAM limit are unchanged. The previous operational wrapper, resource ledger and plan statuses are retained in runtime-recovery-history and included by the amendment analysis manifest. This is recovery in progress, not a completed experiment or performance claim.

## Owner-authorized runtime extension — September 10

The user requested that the runs continue beyond the original deadline. The campaign was safely resumed with a **24-hour total wall ceiling (September 11, 08:59:45 Europe/Berlin)** and **96 CPU-hours**, retaining **four workers and 8 GiB RAM**. This supersedes the original ten-hour/40-CPU-hour ceiling below. It extends time to finish the same matrix; it adds no episodes, replicas, seeds or changes to selection rules.

The original execution source remains `1ce18c8736b6a50773b60b95ec0f11dd4c901028`. The separately committed amendment runner is `355e81d`; the original authorization is unchanged byte-for-byte. Completed blocks and cumulative CPU/wall/peak-memory accounting are preserved, together with interrupted attempts and original status/resource snapshots. Four in-progress blocks restart from their preceding saved stages. The new Python supervisor is PID 44456; all four workers are active. Validation: 743 tests and 11 subtests passed, lint/compilation passed.

This budget amendment follows inspection of preliminary training logs; no evaluation result or selection was available. Disclose the amendment with results. Details and exact resume/analysis commands: `experiments/2026-09-10-task2-rehearsal-mask/runtime-extension.md`. Use the retained `resume_issue124_amendment.py` for subsequent resume/analysis; the old launcher intentionally rejects changed resource limits. No peer approval or merge is implied.

## Active execution — September 10

Julius explicitly overrode the pre-training peer-review gate: "I want you to run it anyways!" This is owner authorization for execution only, not peer approval or merge authorization. The exception is recorded in campaign authorization and each job's metadata.

The four-arm campaign started locally at 08:59:45 Europe/Berlin from immutable commit `1ce18c8736b6a50773b60b95ec0f11dd4c901028`. All four arms were verified running, with four active workers, about 1.2 GiB aggregate RAM and no resource breach. The hard wall deadline is September 10 at 18:59:45 Europe/Berlin. Limits remain four workers, 8 GiB RAM, ten wall hours and 40 CPU-hours. Serial evaluations and automatic analysis follow training within the same budget. No performance result is claimed.

Validation of the explicit exception: 735 tests and 11 subtests passed, lint/compilation/PowerShell syntax passed. CI for the new commit is pending; the preceding implementation's CI passed. Dedicated execution checkout: `issue124-execution`. Outputs: `training_outputs/issue124`. Actual Python supervisor PID at startup: 33336 (Windows launcher PID: 22824). Preserve outputs and do not modify the running checkout.

The protocol below remains the registered scientific design; earlier statements that execution had not started describe preparation and are superseded by this execution record.

## Outcome and scope
Run one bounded Task 2 optimization campaign before the Task 3 handoff. This explicitly supersedes #106's earlier no-further-Task-2 scheduling restriction; it does not extend or reopen #107. Owner: 1BlauNitrox. Prerequisite: #123 / PR #129. Executable protocol: PR #131, branch `experiment/124-task2-rehearsal-mask`, commit `3500468`. Requested reviewer: Waffelmanufaktur. Successor: #125 / PR #130, then #109 / PR #120.

Status: executable prospective protocol; no scientific run started. Julius confirmed all four arms and detached local execution with four workers, 8 GiB aggregate campaign RAM and ten wall hours. The derived CPU ceiling is 40 hours. All nine CI checks pass; peer review remains the launch gate; no further owner compute authorization is needed within this budget.

## Controlled design
All trained arms enable escape-continuation inputs, use uniform replay and identical rewards/network/optimizer/exploration settings. Initialization is the checksum-pinned corrected fresh 26-input Task 1 migration, not a selected #107 replica.

| Arm | Task 1 practice | Legal masking |
| --- | --- | --- |
| A | blocked | off |
| B | interleaved | off |
| C | blocked | on |
| D | interleaved | on |

Hypotheses: interleaving preserves navigation; legal masking with escape inputs improves collection; the combination may differ from the negative #86 result. Zero invalid actions alone is not success.

Five paired replicas per arm, twenty 500-episode blocks: 2,000 coin-heaven, 2,000 loot-crate and 6,000 classic episodes per replica, no opponents. Total 200,000 training episodes. Equal stage counts and scenario-occurrence seed offsets prevent boundary-count and repeated-seed confounding. Interleaving coin blocks 1/6/11/16 changes exposure timing, including exploration levels; it does not promise identical trajectories.

## Evaluation and prospective decision
Use the six checksum-pinned run plans and exact seeds in `experiments/2026-09-10-task2-rehearsal-mask/config.yaml`. The 5,120 evaluation episodes include three scenarios, frozen Task 1/untrained references and deterministic repeats; final/confirmation seeds remain unopened. The untrained reference retains its pinned disabled escape setting with zero added input weights. All trained arms enable the features.

Four primary contrasts: rehearsal B-A and D-C on coin collection; masking C-A and D-B on classic collection. Require >=0.10 improvement and positive 98.75% hierarchical paired-bootstrap lower bound (10,000 resamples, four-test Bonferroni family). B/C require collection/survival non-regression against A; D requires both conditional efficacies and non-regression against A/B/C, with 95% lower bounds strictly above -0.05. Preserve every #107 absolute Task 1/2 and latency/determinism gate.

Select mechanically among eligible treatments, otherwise A, using the registered ranking and median replica. Only final-budget checkpoints qualify. Invalid/incomplete evidence yields no selection; failed absolute gates mean an explicitly exploratory Task 3 predecessor. No performance-based extension or best-seed choice.

## Local parallelism, limits and commands
Detected Windows PC: 16 logical CPUs, about 16 GiB installed RAM. Four concurrent training arms, one worker each; all evaluation serial after the training barrier. The shared monitor includes supervisor usage and retains original CPU/wall/RAM accounting across resume. Startup requires 4 GiB available; runtime aborts below 1 GiB system availability. Only campaign workers are stopped. Keep the computer powered and awake.

From a clean checkout of the exact reviewed PR #131 head:

```powershell
python -m training.run_issue124_campaign --dry-run
.\scripts\start_issue124.ps1 -Python C:\path\to\.venv\Scripts\python.exe -OutputRoot C:\path\to\issue124-output
Get-Content C:\path\to\issue124-output\campaign-status.json
Get-Content C:\path\to\issue124-output\resources.json
python -m training.analyze_issue124_campaign --campaign-root C:\path\to\issue124-output
```

For technical interruption only, inspect previous workers and use the same launcher with `-Resume`. Exhausted ceilings cannot be reset. Keep all outputs until durable evidence is published and reviewed. Ten hours is a hard limit, not a completion guarantee.

## Acceptance and launch gates
- [x] Replay audit and paired stage-seed support implemented in PR #129; review remains required.
- [x] Four-arm scope and exact local compute budget authorized by Julius.
- [x] Plans, prospective decision rules, analyzer and shared monitor committed before compute.
- [x] Local validation: 734 tests + 11 subtests, lint, compile, dry-run, launcher syntax and isolated 14-episode integration smoke pass.
- [x] Detached Windows launch/resume/status/analysis documented; no scientific run yet.
- [x] All nine CI checks pass at `3500468`, including Docker compatibility.
- [ ] Non-author approval on the exact execution revision.
- [ ] All planned results or technical failures retained with durable evidence and mechanical selection.
- [ ] Task 3 handoff records feature dimension, mask mode, provenance and remaining failed gates.

AI assistance: OpenAI Codex implemented and tested the prospective tooling; disclosure in docs/0006-ai-usage.md. Human peer review remains required. No performance claim is made.

**Comment by 1BlauNitrox** [comment]
﻿Verified incomplete outcome at commit 717ec1a: the campaign stopped September 10 at 21:38 Berlin when system-wide free RAM fell below 1 GiB. Completed blocks: A 100/100, B 95/100, C 100/100, D 90/100 (385/400; 192,500 completed training episodes). B/r2/block16 recovery succeeded. B/r5/block16 and D/r5/block11 stopped at the memory guard.

No evaluation ran (0/5,120). The registered amendment-aware analyzer rejects the breached resource record; no scientific selection or Task 2 completion is claimed. Training scores cannot establish a treatment winner. The verified report and compact operational records are committed in experiments/2026-09-10-task2-rehearsal-mask/INTERRUPTION.md and interruption-evidence.json.

At 23:16 only ~1.41 GiB system RAM was available, below the 4 GiB startup requirement. Recovery needs available memory plus a documented technical amendment preserving the breach, original wall clock and cumulative accounting; do not clear the resource flag or bypass the supervisor. Remaining work: 15 blocks plus all evaluation/analysis. Existing 24-hour deadline remains September 11 08:59:45 Berlin. No further training was launched during this inspection.

**Comment by 1BlauNitrox** [comment]
﻿Owner-authorized low-memory recovery started September 10 at 23:26 Berlin after available RAM recovered to 4.31 GiB. Supervisor PID 17268 has two active workers for the remaining B/D training; A/C completed blocks are preserved. The original execution SHA remains 1ce18c8, with unchanged episode/seed/selection rules, cumulative CPU/peak-memory accounting and September 11 08:59:45 Berlin deadline.

Commit b51afb4 adds a tested recovery tool that archives the stopped resource/status/amendment records and its own bytes with hashes before opening the resumed resource segment. It permits only a system-free-memory-floor stop, requires >=4 GiB available RAM and rejects active workers or exhausted budgets. The prior breach remains in the amendment and analysis evidence; safeguards remain active. Validation: 752 tests + 11 subtests, scoped Ruff and compilation passed. No scientific result or completion is claimed yet.

**Comment by 1BlauNitrox** [comment]
﻿September 11 08:10 Berlin verification: campaign stopped again at 00:56:43 because system free RAM fell below 1 GiB. A/B/C: 100 training blocks each; D: 95. Total 395/400 blocks, 197,500 completed training episodes; evaluation 0/5,120. No active workers. Registered analysis-only command rejects the breached resource record, so no scientific selection or Task 2 completion is available.

Current free RAM ~3.37 GiB is below the 4 GiB startup floor. Existing deadline remains 08:59:45 Berlin; five training blocks plus all evaluation/analysis remain. A further recovery requires stable RAM and an explicit budget amendment for work beyond that deadline. No runs restarted and no data deleted.

Committed SECOND-INTERRUPTION.md and second-interruption-evidence.json retain the latest verified operational evidence, preserving the first snapshot. No agent optimization is justified by this campaign yet.

**Comment by 1BlauNitrox** [comment]
﻿Evaluation-only recovery launched September 11 at 08:21 Berlin, supervisor PID 298636. Registered before evaluation in commit 0d7e9e5: A/B/C five completed replicas each; D r1-r4 exploratory; original references, seeds, repeats and metrics; 4,880 evaluation episodes and zero training. All final artifacts are checksum-bound in reduced-evaluation-binding.json.

One serial worker, eight wall hours/eight CPU-hours, 2 GiB aggregate campaign RAM; 2 GiB available startup floor for this single-worker evaluation, unchanged 1 GiB runtime availability floor. New segment deadline approximately 16:21 Berlin. Original stopped training ledger/attempts remain untouched. D/r5 is excluded solely for technical incompletion; no full-matrix completion claim. A/B/C retain original gates and conservative multiplicity correction.

Validation: 758 tests and 11 subtests, scoped Ruff and compilation passed. Evaluation outputs are in training_outputs/issue124-reduced-evaluation; status.json distinguishes evaluation completion from scientific analysis, which remains pending. No performance claim yet. Keep the preparation and immutable execution checkouts unchanged during execution.

**Comment by 1BlauNitrox** [comment]
﻿Owner-requested throughput-based evaluation extension applied at 08:48 Berlin, commit bd4d91f. Effective limits: 24 wall hours from the original 08:21 start (deadline September 12 08:21), 32 cumulative CPU-hours, unchanged 2 GiB campaign RAM / one worker / 1 GiB system availability floor. Original registration and authorization are untouched. runtime-extension.json binds the original/replacement runners and archived ledgers/statuses; cumulative accounting includes conservative 12-second CPU padding for handover. Completed episodes retained; incomplete in-flight evaluation is retried with its recorded seeds.

Replacement supervisor PID 312636 is active. Windows idle-sleep prevention is requested for the supervisor lifetime without changing permanent power settings. No guarantee against power loss or system memory exhaustion. Validation: 759 tests + 11 subtests, scoped Ruff passed. No agent-performance outcomes were used to decide the extension.

Task 3 is not blocked for preparation: PR 130 has all CI checks green at 797eae7 but requires fresh non-author implementation review; PR 120 is approved. Final selected-parent binding remains dependent on evaluated Task 2 gates.

**Comment by 1BlauNitrox** [comment]
﻿## Result
Refs #124. Record the completed **reduced** Task 2 rehearsal/masking evaluation: 19 trained replicas, 4,880 laptop evaluation episodes, exact final checkpoint provenance, deterministic repeats and passing latency/resource checks. D/r5 training is incomplete; D remains exploratory and the original full-matrix issue stays open.

Rehearsal B-A reduces coin-heaven collection by 16.99 percentage points (corrected 98.75% CI -26.82 to -6.82). Masking C-A improves classic collection by 5.33 points (corrected CI -7.86 to +17.33), failing the registered 10-point/confidence adoption criteria. C passes Task 2-specific gates but fails Task 1 retention. No candidate passes cumulative completion gates. Mechanical fallback A/r5 remains an exploratory predecessor; no agent checkpoint/default was replaced.

## Reviewable evidence
- `experiments/2026-09-10-task2-rehearsal-mask/RESULTS.md`: interpretation, limitations, next decisions and exact reproduction commands.
- `laptop-results/evidence.json.gz`: compact lossless observations, metadata, original training provenance, migration/resource records and hashes.
- `episodes.csv`, `summary.csv`, `result.json`: all primary/repeat observations and recomputable decisions/uncertainty.
- https://github.com/1BlauNitrox/mle-final-project/releases/tag/issue124-reduced-evidence-v1 : retrievable 3,852,548-byte checkpoint archive, 21 artifacts and manifest. SHA-256 `dbdf7d81fb61bef9de6882c4fcfa59610c0ed9957da1aa345210599b9b327e59`.

The reduced analyzer verifies complete A/B/C five-replica conditions; D contrasts use matched r1-r4 only and cannot select D. Laptop results use Python 3.13.15 and pinned source 1ce18c8. Old PC observations are historical, not pooled. The original evaluation budget clock and CPU usage are retained.

## Integration and validation
Current main 08ca233 was merged. A later #139 seed-number overlap is explicitly disclosed: historical validation uses the checksum-pinned registration inventory, while default new-launch validation still rejects the live collision. No experiment seeds were changed.

- Full local suite: 861 tests and 11 subtests passed.
- Inventory LF/checksum follow-up: 22 focused tests passed.
- Changed-code Ruff, compilation and diff checks passed.
- Six reduced-analysis regression tests cover reproduction and rejection of missing jobs, changed seeds, wrong artifacts, reset CPU accounting and mismatched repeats.
- Current-head CI pending.

## Scope / Definition of Done
- [x] Completed reduced-result record and implementation validation.
- [x] Durable per-observation evidence, artifact retrieval and exact analysis command.
- [x] Negative results, incomplete D/r5, Task 1 failures and unsupported completion claims explicitly retained.
- [x] Agent card and AI disclosure updated.
- [x] Local validation passed.
- [ ] Current-head CI passed.
- [ ] Fresh non-author approval and resolved review conversations.
- [ ] Original full-matrix #124 and cumulative DQN Task 2 completion (not claimed).

Codex assisted verification, analysis tooling, tests and documentation; disclosed in docs/0006-ai-usage.md. Human scientific/code review remains required. No new training or evaluation was launched during analysis.
